raft/server.cc

```diff
@@ -921,8 +921,8 @@ future<> server_impl::io_fiber(index_t last_stable) {
                 for (const auto& addr: rpc_diff.leaving) {
                     abort_snapshot_transfer(addr.id);
                     remove_from_rpc_config(addr);
-                    _rpc->remove_server(addr.id);
                 }
+                _rpc->on_configuration_change({}, rpc_diff.leaving);
```

Comment:
@gusev-p please check it out. @kbr- has a good point here. We don't co_await for send_message return, in fact, send_message is not returning a future. send_append_entries_reply() takes a gate before accessing raft address map, so the address resolution may happen after a yield. 
We either need to make sure we wait for all send_message() to resolve here, or move get_inet_asddress() in send_append_entries_reply() & friends to a place before taking the gate, or use the approach taken in this patch - never directly delete addresses from address map, only mark them as expiring.

The key takeaway is that @kbr- apparently identified the cause of the bug you are invesitgating.

Reply:
@kbr- as to your comment, I'd like to know what Gleb thinks. My initial reaction is that we need to move address resolution to before-the-yield, since this is how this code was written initially, and shutdown gate was added as an afterthought.

Reply:
@gleb-cloudius the alternative is what kamil suggests - merge notifications about joining and leaving nodes into a single call and *always* merely mark the nodes as expiring when they are leavinig, not erase them from the address map (this is what this patch already implements).

Reply:
While @kbr-  is right that it is not necessary to send a messages to the removed node (the protocol should be correct even when messages are dropped) it may help for the removed node to learn about it been removed and to stop trying to disturb a cluster (or even shutdown itself earlier).

About the solution he propose to make an entry expiring instead of removing, it will work but it means that our raft implementation will depend on an internals of an rpc implementation. My be we should consider resolving before yield. Note that the yield is hard to hit, but it definitely looks like the bug we hit in longevity.

Reply:
>  it will work but it means that our raft implementation will depend on an internals of an rpc implementation

Or we can put an explicit requirement on our rpc interface definition. Then it becomes a contract, not implementation detail.

Reply:
> > it will work but it means that our raft implementation will depend on an internals of an rpc implementation
> 
> Or we can put an explicit requirement on our rpc interface definition. Then it becomes a contract, not implementation detail.

But this will be a strange contract: "keep an entry for a while after it has been deleted". How you define "for a while"?

Reply:
With the most recent changes where the mapping resides completely outside raft we could say that "raft doesn't care" how the mapping is maintained, it just pushes the FSM changes out to RPC which can then do whatever. But having it in two calls makes it more flexible to the RPC implementation, so why not keep it?

Reply:
@gleb-cloudius

Reply:
Raft does care about a mapping lifetime though. I do not see how you can tel raft entirely does not care. The assumption out current code as is that messages that are sent before address removal will be able to use a mapping that was there before.

Reply:
So this is exactly what I propose - preserve the API and calling convention, and consider fixing the issue of a yield in raft_rpc implementation.

Reply:
Yield is in the raft code, but otherwise it looks like we are on the same page then.